### PR TITLE
Support drawing mid-circuit measurement statistics with `qml.draw`

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
This PR adds support for drawing the collection of mid-circuit measurement statistics with `qml.draw`. An example is shown below:
```python
def circ():
    qml.Hadamard(0)
    m0 = qml.measure(0)
    qml.Hadamard(1)
    m1 = qml.measure(1)
    qml.Hadamard(2)
    m2 = qml.measure(2)
    qml.Hadamard(3)
    qml.measure(3)
    qml.cond(m0, qml.PauliX)(0)
    qml.cond(m0 & m1, qml.PauliY)(1)
    return qml.expval(m2), qml.sample([m1, m2])

print(qml.draw(circ)())
```
```
0: ──H──┤↗├──────────────────────────X────┤                    
1: ──────║───H──┤↗├──────────────────║──Y─┤                    
2: ──────║───────║───H──┤↗├──────────║──║─┤                    
3: ──────║───────║───────║───H──┤↗├──║──║─┤                    
         ╚═══════║═══════║═══════════╩══╣                      
                 ╚═══════║══════════════╩═╡        ╭Sample[MCM]
                         ╚════════════════╡  <MCM> ╰Sample[MCM]
```

**Description of the Change:**

* Refactor `tape_text` to only run through one loop containing all layers, not separate loops for op layers and measurement layers. The logic inside the loop has been adjusted accordingly. I also added big block comments to make the separation between different steps more obvious.
* Add helpers to `tape_text.py` to add grouping symbols and labels for mid-circuit measurement statistics correctly.
* Update `drawable_layers` to find layers for mid-circuit measurement statistics correctly. This included adding a new helper to recursively find layers for terminal measurements that use measurement values.
* The bit map is now initialized with `None` values before everything else, and this is used to keep track of which mid-circuit measurements are used. Mostly a refactor to tidy up the code.
* Remove code from `draw_mpl` that deferred measurements unconditionally for qnodes with mid-circuit measurements. This code was removed from `qml.draw` in #4901 and `qml.draw_mpl` is now updated as well.
* The example in the `iterative_qpe` docstring was written before we had advanced mid-circuit measurement support and drawing capabilities, so I updated the example as well.

**Benefits:**
* We can draw mid-circuit measurement statistics!
* Code related to the text drawer is slightly more readable.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/drawer/draw.py [hunk 1]
```diff
@@ -573,16 +573,7 @@ def wrapper(*args, **kwargs_qnode):
...
-                        isinstance(op, qml.measurements.MidMeasureMP)
-                        for op in qnode.tape.operations
-                    ):
-                        tapes, _ = qml.defer_measurements(qnode.tape, device=qnode.device)
-                    else:
-                        tapes = [qnode.tape]
-
-                    tapes, _ = program(tapes)
-                    tape = tapes[0]
+                    [tape], _ = program([qnode.tape])
```
Comment (albi3ro) [inline pennylane/drawer/draw.py:576]: We probably need to do this for `draw_mpl` as well.

Comment (mudit2812) [reply pennylane/drawer/draw.py:576]: This is for `draw_mpl`. I updated it in `draw` in #4901.

**Comment by mudit2812** [comment]
[sc-49830]

File: tests/drawer/test_drawable_layers.py [hunk 2]
```diff
@@ -140,22 +190,27 @@ def test_all_wires_measurement(self, measurement):
...
         def teleport(state):
             qml.QubitStateVector(state, wires=["A"])
             qml.Hadamard(wires="a")
             qml.CNOT(wires=["a", "B"])
             qml.CNOT(wires=["A", "a"])
             qml.Hadamard(wires="A")
-            m0 = qml.measure("A")
-            m1 = qml.measure("a")
+            qml.apply(mp0)
+            qml.apply(mp1)
```
Comment (timmysilv) [inline tests/drawer/test_drawable_layers.py:205]: why was this change necessary?

**Review by timmysilv — COMMENTED** [review]
nice stuff!

File: pennylane/ops/functions/iterative_qpe.py [hunk 3]
```diff
@@ -39,34 +39,39 @@ def iterative_qpe(base, ancilla, iters):
...
         >>> print(circuit())
-        [array([0, 0, 0, 0, 0]), array([1, 0, 0, 0, 0]), array([0, 1, 1, 1, 1])]
+        [[0 0 1]
+         [0 0 1]
+         [0 0 1]
+         [1 1 1]
+         [0 0 1]]
 
-    The output is an array of size ``(number of iterations, number of shots)``.
+    The output is an array of size ``(number of shots, number of iterations)``.
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/ops/functions/iterative_qpe.py:64]: It was correct earlier because Guillermo was returning a list with `qml.sample`s for each iteration

[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/ops/functions/iterative_qpe.py:64]: when did this change?

File: tests/drawer/test_drawable_layers.py (hunk 2 above)
Comment (mudit2812) [reply tests/drawer/test_drawable_layers.py:205]: So that the `MidMeasureMP`s that I put in the bit map have the same ID as the original `MidMeasureMP`s.

File: pennylane/ops/functions/iterative_qpe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/ops/functions/iterative_qpe.py:64]: This PR. It's just that Guillermo was returning `[qml.sample(m) for m in measurements]` before because we couldn't collect statistics on multiple mid-circuit measurements. Now that's possible, so I changed it to `qml.sample(measurements)` instead. In both cases, the shape is correct, but the original way was just because of constraints which no longer exist.

File: pennylane/drawer/tape_text.py [hunk 4]
```diff
@@ -160,13 +168,51 @@ def _add_mid_measure_op(op, layer_str, config):
...
+
+
+def _add_cwire_measurement(m, layer_str, config):
+    """Updates ``layer_str`` with the ``m`` measurement when it it used
+    for collecting mid-circuit measurement statistics."""
+    mcms = [v.measurements[0] for v in m.mv] if isinstance(m.mv, list) else m.mv.measurements
+    layer_str = _add_cwire_measurement_grouping_symbols(mcms, layer_str, config)
+
+    mv_label = "MCM"
+    meas_label = measurement_label_map[m.return_type](mv_label)
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/drawer/tape_text.py:201]: That's a good point. However, I'd rather keep it as is because this label was approved by @trbromley . Since he's on vacation. We discussed potentially revisiting this in the future and looking into better labels for composite measurement values, so it could change later (or not) 😄

File: pennylane/drawer/drawable_layers.py [hunk 5]
```diff
@@ -115,41 +147,51 @@ def drawable_layers(ops, wire_map=None):
...
+            op_occupied_wires = set()
+            mapped_cwires = (
+                [bit_map[m.measurements[0]] for m in mv]
+                if isinstance(mv, list)
+                else [bit_map[m] for m in mv.measurements]
+            )
+            op_occupied_cwires = set(range(min(mapped_cwires), max(mapped_cwires) + 1))
+            op_layer = _recursive_find_mcm_stats_layer(
+                max_layer, op_occupied_cwires, used_cwires_per_layer, bit_map
+            )
```
Comment (albi3ro) [inline pennylane/drawer/drawable_layers.py:183]: Can we switch the order of these two branches? It's often easier to read if we can control on the positive (`condition`) then the negative (`not condition`) . The `not` gives you one more piece of information to carry around in your active memory.

Also, even though `getattr(op, "mv", None)` does indeed get the job done, it might not be as readable to someone less familiar with mid circuit measurements.  Maybe instead `isinstance(op, MeasurementProcess) and op.mv`?

File: pennylane/drawer/drawable_layers.py [hunk 6]
```diff
@@ -50,12 +50,46 @@ def _recursive_find_layer(layer_to_check, op_occupied_wires, occupied_wires_per_
...
+        int: layer to place measurement process in
+    """
+
+    if used_mcms_per_layer[layer_to_check]:
+        stat_cwires = [bit_map[m] for m in stat_mcms]
+        op_occupied_cwires = set(range(min(stat_cwires), max(stat_cwires) + 1))
+        layer_cwires = [bit_map[m] for m in used_mcms_per_layer[layer_to_check]]
+        layer_occupied_cwires = set(range(min(layer_cwires), max(layer_cwires) + 1))
+
+        if op_occupied_cwires & layer_occupied_cwires:
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/drawer/drawable_layers.py]: This implementation was a bit over-engineered. I've updated it now and it should be slightly easier to understand.

File: pennylane/ops/functions/iterative_qpe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/ops/functions/iterative_qpe.py:64]: oh I'm being dumb. you didn't change iqpe, but you did change the QNode in the docstring. lgtm

File: pennylane/drawer/tape_text.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/drawer/tape_text.py:201]: very well

**Review by albi3ro — APPROVED** [review]
👍 Looks great :)

**Review by timmysilv — APPROVED** [review]
✏️ 🔬 :electron: 🎉
